eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,354 @@
+	err = router.CreateBitbucketWebhook(ctx, bitbucketConfig)
+
+	if err != nil {
+		logger.Errorw("failed to create/update Bitbucket webhook", zap.Error(err))
+	}
+
+	go func() {
+		// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
+		// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from Bitbucket.
+		// This is a workround to mitigate the race conditions.
+		logger.Info("starting bitbucket hooks manager daemon")
+		ticker := time.NewTicker(60 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Info("exiting bitbucket hooks manager daemon")
+				return
+			case <-ticker.C:
+				err := router.CreateBitbucketWebhook(ctx, bitbucketConfig)
+
+				if err != nil {
+					logger.Errorw("failed to create/update Bitbucket webhook", zap.Error(err))
+				}
+			}
+		}
+	}()
+
+	return webhook.ManageRoute(ctx, router, controller, dispatch)
+}
+
+func (router *Router) CreateBitbucketWebhook(ctx context.Context, bitbucketConfig *bitbucketv1.Configuration) error {
+	bitbucketserverEventSource := router.bitbucketserverEventSource
+	route := router.route
+
+	logger := route.Logger.With(
+		logging.LabelEndpoint, route.Context.Endpoint,
+		logging.LabelPort, route.Context.Port,
+		logging.LabelHTTPMethod, route.Context.Method,
+		"project-key", bitbucketserverEventSource.ProjectKey,
+		"repository-slug", bitbucketserverEventSource.RepositorySlug,
+		"base-url", bitbucketserverEventSource.BitbucketServerBaseURL,
+	)
+
+	bitbucketClient := bitbucketv1.NewAPIClient(ctx, bitbucketConfig)
+
+	formattedURL := common.FormattedURL(bitbucketserverEventSource.Webhook.URL, bitbucketserverEventSource.Webhook.Endpoint)
+
+	apiResponse, err := bitbucketClient.DefaultApi.FindWebhooks(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to list existing hooks to check for duplicates for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	hooks, err := bitbucketv1.GetWebhooksResponse(apiResponse)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert the list of webhooks for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	var existingHook bitbucketv1.Webhook
+	isAlreadyExists := false
+
+	for _, hook := range hooks {
+		if hook.Url == formattedURL {
+			isAlreadyExists = true
+			existingHook = hook
+			router.hookID = hook.ID
+			break
+		}
+	}
+
+	logger.Info("retrieving the webhook secret...")
+	webhookSecret, err := router.getWebhookSecret(bitbucketserverEventSource.WebhookSecret)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver webhook secret. err: %+v", err)
+	}
+
+	newHook := bitbucketv1.Webhook{
+		Name:          "Argo Events",
```

Comment:
Does duplicate name matter?

Reply:
It doesn't. Bitbucket Server doesn't care. Matching (to find the existing hook) is based on the Url and updating/deleting is done by Id. I don't like hard coded names like these and I tried to find another way of naming the hook so you can relate it to a specific EventSource, but I failed to find anything that I could use. If you have any suggestion for a better name, please let me know. Otherwise I'll just keep it like this.

Reply:
I'm ok with it as long as bitbucket doesn't care.